src/AzureExtension/Helpers/AzureRepositoryHierarchy.cs

```diff
@@ -72,17 +72,21 @@ public async Task<List<TeamProjectReference>> GetProjectsAsync(Organization orga
         // Makes sure _organizationsAndProjects has all organizations.
         await Task.Run(() => GetOrganizations());
 
-        lock (_getProjectsLock)
-        {
-            _organizationsAndProjects.TryGetValue(organization, out var projects);
+        _organizationsAndProjects.TryGetValue(organization, out var projects);
 
-            if (projects == null || projects.Count == 0)
+        if (projects == null || projects.Count == 0)
+        {
+            lock (_getProjectsLock)
             {
-                _organizationsAndProjects[organization] = QueryForProjects(organization);
+                // Double check against race conditions.
+                if (projects == null || projects.Count == 0)
+                {
+                    _organizationsAndProjects[organization] = QueryForProjects(organization);
+                }
             }
-
-            return _organizationsAndProjects[organization];
         }

```

Comment:
I don't understand this locking code - is it right?

I see only one lock on '_getProjectsLock' but many users of '_organizationsAndProjects'.

And there's a double-checked lock - which is always super fishy (at least if the caller requires the returned information to be correct.)